src/Catalog/CommitCollector.cs

```diff
@@ -56,10 +57,25 @@ public CommitCollector(Uri index, Func<HttpMessageHandler> handlerFunc = null)
                         batch.CommitTimeStamp == lastBatch.CommitTimeStamp, 
                         cancellationToken);
 
-                    front.Value = batch.CommitTimeStamp;
-                    await front.Save(cancellationToken);
+                    // Only commit if this is the last batch that is going to be processed or if the commit timestamp
+                    // has changed from the previous batch. This is important because if two batches have the same
+                    // commit timestamp but processing the second fails, we should not progress the cursor forward.
+                    if (ReferenceEquals(batch, lastBatch))
+                    {
+                        front.Value = batch.CommitTimeStamp;
+                        await front.Save(cancellationToken);
+                        Trace.TraceInformation("CommitCatalog.Fetch front.Value saved due to last batch: {0}", front);
+                    }
+                    else if (previousCommitTimeStamp.HasValue && previousCommitTimeStamp != batch.CommitTimeStamp)
```

Comment:
Is it safe to reprocess a successfully processed batch?

Since timestamp alone is inadequate precision for the cursor, what should the cursor be based on?